### PR TITLE
149 server sends mlgym event messages in json format to client

### PR DESCRIPTION
**Pull request by OsamaMSoliman**
Waiting for PR #147 , before marking this PR as ready, and afterwards create a new PR for #152

### CODE REVIEW

File: src/ml_board/backend/dummy_server/dummy_server.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
+####################################################################################################
+# imports
```
Comment (le1nux) [inline src/ml_board/backend/dummy_server/dummy_server.py:2]: comment stating the obvious

File: src/ml_board/backend/dummy_server/dummy_server.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import json
+import sys
+from collections import defaultdict
+from typing import List
+
+from flask import Flask, copy_current_request_context, render_template, request
+from flask_socketio import SocketIO, disconnect, emit, join_room, rooms
+
+####################################################################################################
```
Comment (le1nux) [inline src/ml_board/backend/dummy_server/dummy_server.py:11]: I think this is legacy and can be removed

File: src/ml_board/backend/dummy_server/dummy_server.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+# TODO: ASK Max what's that for?
+# class EventSubscriberIF:
+#     def callback(self):
+#         raise NotImplementedError
+
+
+####################################################################################################
+# Flask HTTP Server
+def CreateFlaskServer() -> Flask:
```
Comment (le1nux) [inline src/ml_board/backend/dummy_server/dummy_server.py:20]: do we need this? Also function does not follow PEP-8

File: src/ml_board/backend/dummy_server/dummy_server.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+# def _api_status() -> dict:
+#     client_sids = ws.client_sids
+#     room_key_to_sid = defaultdict(list)
+#     for client_sid in client_sids:
+#         room_keys = rooms(client_sid, "/")
+#         for room_key in room_keys:
+#             room_key_to_sid[room_key].append(client_sid)
+#     return {"clients": client_sids, "rooms": room_key_to_sid}
+
+# TODO: RESTful APIs could be added here!
```
Comment (le1nux) [inline src/ml_board/backend/dummy_server/dummy_server.py:51]: this should be another PR and be documented in an issue rather than in code. But the idea is good :+1:

File: src/ml_board/backend/dummy_server/dummy_server.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+    return _app
+
+
+def _index() -> str:
+    # return render_template('index.html', async_mode=ws._socketio.async_mode)
+    return "Index Page"
+
+
+# def _status() -> str:
```
Comment (le1nux) [inline src/ml_board/backend/dummy_server/dummy_server.py:38]: lots of legacy code that needs to be removed before merging

File: src/ml_board/backend/dummy_server/dummy_server.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+class WebSocketWrapper:
+    def __init__(self, flask_app: Flask, host: str, port: int, async_mode: str, message_delay: int, log_file_path: str, how_many_lines: int):
+        # array of connected client session ids
+        self._client_sids = []
+        self._message_delay = message_delay
+
+        # create the websocket
+        self._socketio = SocketIO(flask_app,
+                                  async_mode=async_mode,
+                                  cors_allowed_origins=["http://localhost:3000", "http://localhost:7000", "http://localhost:8080",
```
Comment (le1nux) [inline src/ml_board/backend/dummy_server/dummy_server.py:67]: can we pass in the `cors_allowed_origins` as CMD params?

File: src/ml_board/backend/websocket_api/websocket_server_dummy.py [hunk 7]
```diff
@@ -12,7 +15,7 @@ def callback(self):
 
 class WebSocketServer:
```
Comment (le1nux) [inline src/ml_board/backend/websocket_api/websocket_server_dummy.py:16]: Do we still need this now that we have dumm_server.py?

**Review by le1nux — CHANGES_REQUESTED** [review]
I think it still needs some cleaning before merging

File: src/ml_board/frontend/Dashboard/src/redux/table/tableSlice.ts [hunk 8]
```diff
@@ -4,7 +4,7 @@ import { RootState } from "../store";
 
 // Row = JobStatusPayload + ExperimentStatusPayload
 export interface Row {
-    // Job
+    // // Job
```
[the remark this answers is not in the document]
Comment (OsamaMSoliman) [reply src/ml_board/frontend/Dashboard/src/redux/table/tableSlice.ts:7]: No no this is intentional as a comment to show the keys of the Row object (ideally)!!

File: src/ml_board/backend/websocket_api/websocket_server_dummy.py (hunk 7 above)
Comment (OsamaMSoliman) [reply src/ml_board/backend/websocket_api/websocket_server_dummy.py:16]: I personally don't use this anymore!

File: src/ml_board/backend/dummy_server/dummy_server.py (hunk 5 above)
Comment (OsamaMSoliman) [reply src/ml_board/backend/dummy_server/dummy_server.py:38]: I'll delete **_status** but **_api_status** was in the old code but not used and I'm not sure how was it intended to be used?!!?! 🤷

File: src/ml_board/backend/dummy_server/dummy_server.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        # self._emit_server_log_message(f"Client {client_id} joined rooms: {rooms()}")
+
+        # send the history of log messages to the client
+        if "mlgym_event_subscribers" in data['rooms']:
+            self._socketio.start_background_task(_send_event_history_to_client)
+
+    def _on_leave(self):
+        self._client_sids.remove(request.sid)
+        # TODO  leave all rooms
+        # leave_room(message['room'])
```
[the remark this answers is not in the document]
Comment (OsamaMSoliman) [reply src/ml_board/backend/dummy_server/dummy_server.py:127]: I didn't change it, it was also committed in the old code and I actually wanted to ask about it!

uncommenting it leads to errors `leave_room` and `message` are undefined!

File: src/ml_board/backend/dummy_server/dummy_server.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        print("how_many_lines could be -1 for the entire file!")
+        sys.exit(0)
+    argvs = dict(zip(arg_keys, sys.argv))
+
+    try:
+        # run the Server
+        WebSocketWrapper(
+            flask_app=CreateFlaskServer(),
+            host="localhost",
+            port=7000,
```
[the remark this answers is not in the document]
Comment (OsamaMSoliman) [reply src/ml_board/backend/dummy_server/dummy_server.py:155]: again sure but I don't want to over complicate it in the dummy server! 😅

File: src/ml_board/backend/dummy_server/dummy_server.py [hunk 11]
```diff
@@ -136,27 +122,29 @@ def _on_leave(self):
...
 
-    # make sure the user input all arg_keys
-    input_len = len(sys.argv)
-    arg_keys = ['program_name', 'log_file_path', 'how_many_lines']
-    if input_len != len(arg_keys):
-        print(f"missing args: {arg_keys[input_len:]}")
-        print("how_many_lines could be -1 for the entire file!")
+    sys.argv.pop(0)
+    args = dict([arg.split('=') for arg in sys.argv])
+    print(f"{args}\n{'Valid parameters ✔' if all(arg in params for arg in args) else 'Invalid parameters ✗'}")
```
Comment (le1nux) [inline src/ml_board/backend/dummy_server/dummy_server.py]: no special characters in code only ascii

**Review by le1nux — APPROVED** [review]

File: src/ml_board/backend/dummy_server/dummy_server.py (hunk 11 above)
Comment (OsamaMSoliman) [reply src/ml_board/backend/dummy_server/dummy_server.py]: Done, removed them!

File: src/ml_board/frontend/Dashboard/src/redux/table/tableSlice.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (OsamaMSoliman) [reply src/ml_board/frontend/Dashboard/src/redux/table/tableSlice.ts:7]: I did actually that (for example: `job_id?:string;` or generally `[key:string]:string;`) but then I faced errors later and that's why i removed them again!

**Comment by OsamaMSoliman** [comment]
This PR also fixes the same issue that #158 was trying to fix!
